bson/src/main/org/bson/codecs/DocumentCodec.java

```diff
@@ -216,10 +229,29 @@ private Object readValue(final BsonReader reader, final DecoderContext decoderCo
             return null;
         } else if (bsonType == BsonType.ARRAY) {
             return readList(reader, decoderContext);
-        } else if (bsonType == BsonType.BINARY && BsonBinarySubType.isUuid(reader.peekBinarySubType()) && reader.peekBinarySize() == 16) {
-            return registry.get(UUID.class).decode(reader, decoderContext);
+        } else {
+            Codec<?> codec = bsonTypeCodecMap.get(bsonType);
+            
+            if (bsonType == BsonType.BINARY && reader.peekBinarySize() == 16) {
+                switch (reader.peekBinarySubType()) {
+                    case 3:
+                        if (uuidRepresentation == UuidRepresentation.JAVA_LEGACY
+                                || uuidRepresentation == UuidRepresentation.C_SHARP_LEGACY
+                                || uuidRepresentation == UuidRepresentation.PYTHON_LEGACY) {
+                            codec = registry.get(UUID.class);
+                        }
+                        break;
+                    case 4:
+                        if (uuidRepresentation == UuidRepresentation.JAVA_LEGACY || uuidRepresentation == UuidRepresentation.STANDARD) {
```

Comment:
Yeah, also weird.  If you look at the current behavior, all subtype 3 and subtype 4 binary value are converted to UUID, and in a minor release that behavior has to be preserved.  Since the default uuidRepresentation is JAVA_LEGACY, then the way to preserver that behavior is to convert to UUID in this case.

In the 4.0 breaking release, that part of the conditional will be dropped, and then subtype 4 binaries will only be converted to UUID when the representation is STANDARD

Reply:
To preserve the current behavior don't you also need to decode subtype 4 to UUID when the representation is PYTHON_LEGACY and C_SHARP_LEGACY? (Basically any representation besides UNSPECIFIED.)

Reply:
Similar to above comment.  There is no way to set the representation now, so this is new behavior.  Since effectively the representation is currently hardcoded to JAVA_LEGACY, I did include JAVA_LEGACY in the conditional to preserve backwards compatibility.

Reply:
I see. Thanks for explaining.